vm/src/bytesinner.rs

```diff
@@ -38,17 +41,14 @@ impl From<Vec<u8>> for PyBytesInner {
 
 impl TryFromObject for PyBytesInner {
     fn try_from_object(vm: &VirtualMachine, obj: PyObjectRef) -> PyResult<Self> {
+        if let Ok(zelf) = try_bytes_like(vm, &obj, |bytes| Self::from(bytes.to_vec())) {
+            return Ok(zelf);
+        }
+
         match_class!(match obj {
-            i @ PyBytes => Ok(PyBytesInner {
-                elements: i.borrow_value().to_vec()
-            }),
-            j @ PyByteArray => Ok(PyBytesInner {
-                elements: j.borrow_value().elements.to_vec()
-            }),
-            k @ PyMemoryView => Ok(PyBytesInner {
-                elements: k.try_bytes(|v| v.to_vec()).unwrap()
-            }),
+            // TODO: generic way from &[PyObjectRef]
             l @ PyList => l.to_byte_inner(vm),
+            t @ PyTuple => t.to_bytes_inner(vm),
             obj => {
                 let iter = vm.get_method_or_type_error(obj.clone(), "__iter__", || {
                     format!("a bytes-like object is required, not {}", obj.class())
```

Comment:
I think we could have a trait for the objects that can be borrowed as &[PyObjectRef], so we can have a generic and efficiency way to iter.